pkg/catalog/announcement_handlers.go

```diff
@@ -1,52 +1,56 @@
 package catalog
 
 import (
-	"errors"
-
-	"k8s.io/apimachinery/pkg/types"
+	v1 "k8s.io/api/core/v1"
 
 	"github.com/openservicemesh/osm/pkg/announcements"
 	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/kubernetes/events"
 )
 
-var errEventNotHandled = errors.New("event not handled")
-
-// releaseCertificate is an Announcement handler, which on receiving a PodDeleted event
-// it releases the xDS certificate for the Envoy for that Pod.
-func (mc *MeshCatalog) releaseCertificate(ann announcements.Announcement) error {
-	whatWeGot := ann.Type
-	whatWeCanHandle := announcements.PodDeleted
-	if whatWeCanHandle != whatWeGot {
-		log.Error().Msgf("releaseCertificate function received an announcement with type %s; it can only handle %s", whatWeGot, whatWeCanHandle)
-		return errEventNotHandled
-	}
-
-	if podUID, ok := ann.ReferencedObjectID.(types.UID); ok {
-		if podIface, ok := mc.podUIDToCN.Load(podUID); ok {
-			endpointCN := podIface.(certificate.CommonName)
-			log.Warn().Msgf("Pod with UID %s found in Mesh Catalog; Releasing certificate %s", podUID, endpointCN)
-			mc.certManager.ReleaseCertificate(endpointCN)
-		} else {
-			log.Warn().Msgf("Pod with UID %s not found in Mesh Catalog", podUID)
+// releaseCertificateHandler releases certificates based on podDelete events
+func (mc *MeshCatalog) releaseCertificateHandler() chan struct{} {
+	podDeleteSubscription := events.GetPubSubInstance().Subscribe(announcements.PodDeleted)
+	stop := make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-stop:
+				return
+			case podDeletedMsg := <-podDeleteSubscription:
+				psubMessage, castOk := podDeletedMsg.(events.PubSubMessage)
+				if !castOk {
+					log.Error().Msgf("Error casting PubSubMessage: %v", psubMessage)
+					continue
+				}
+
+				// guaranteed can only be a PodDeleted event
+				deletedPodObj, castOk := psubMessage.OldObj.(*v1.Pod)
+				if !castOk {
+					log.Error().Msgf("Failed to cast to *v1.Pod: %v", psubMessage.OldObj)
+					continue
+				}
+
+				podUID := deletedPodObj.GetObjectMeta().GetUID()
+				if podIface, ok := mc.podUIDToCN.Load(podUID); ok {
+					endpointCN := podIface.(certificate.CommonName)
+					log.Warn().Msgf("Pod with UID %s found in Mesh Catalog; Releasing certificate %s", podUID, endpointCN)
+					mc.certManager.ReleaseCertificate(endpointCN)
+
+					// Request a broadcast update, just for security.
+					// Dispatcher code also handles PodDelete, so probably the two will get coalesced.
+					events.GetPubSubInstance().Publish(events.PubSubMessage{
+						AnnouncementType: announcements.ScheduleBroadcastUpdate,
+						NewObj:           nil,
+						OldObj:           nil,
+					})
+				} else {
+					log.Warn().Msgf("Pod with UID %s not found in Mesh Catalog", podUID)
+				}
```

Comment:
This here would be a good candidate to carve out into a handler for the particular message type.

Reply:
Will reach you offline to get more insight on this.